source/brailleTables.py

```diff
@@ -82,325 +97,470 @@ def listTables():
+						manifest = CustomTablesManifest(file_)
+						_loadCustomTablesConfig(manifest.dict())
+				except:  # noqa: E722
+					log.exception(
+						"Error while applying custom braille tables config: {}"
+						"".format(path)
+					)
+
+
+def terminate():
+	global tablesDirs, _tables
+	tablesDirs = [TABLES_DIR]
+	_tables = {}
+
+
+def _getCustomTablesDirs():
+	"""Retrieve the custom braille tables directories.
+	
+	These are "brailleTables" sub-directories searched after in order in:
+	 - The scratchpad directory, if enabled.
+	 - The user configuration directory.
+	 - Every running addon.
+	"""
+	import os
+
+	def candidates():
+		import config
+		import globalVars
+		if (
+			not globalVars.appArgs.secure
+			and config.conf['development']['enableScratchpadDir']
+		):
+			yield os.path.join(config.getScratchpadDir(), "brailleTables")
+		import addonHandler
+		for addon in addonHandler.getRunningAddons():
+			yield os.path.join(addon.path, "brailleTables")
+
+	return [dir_ for dir_ in candidates() if os.path.isdir(dir_)]
+
+
+class CustomTablesManifest(ConfigObj):
+	"""Information regarding one or more custom braille tables.
+	"""
+	configspec = ConfigObj(StringIO(
+		"""# NVDA Custom Braille Tables Manifest ConfigObj specification
+# Table file name (not the full path)
+[__many__]
+	contracted = boolean(default=false)
+	input = boolean(default=true)
+	output = boolean(default=true)
+	[[displayName]]
+		# The key is a locale code (eg. "de", "pt_BR", ...).
+		# The value is the displayName for that locale.
+		__many__ = string()"""
+	))
+
+	def __init__(self, file_: typing.BinaryIO):
+		""" Initializes an L{CustomTablesManifest} instance from a file.
+		@param file_: The manifest file-like object, opened in binary mode.
+		"""
+		super(CustomTablesManifest, self).__init__(
+			file_,
+			configspec=self.configspec,
+			encoding='utf-8',
+			default_encoding='utf-8'
+		)
+
+
+def _loadCustomTablesConfig(data: dict):
+	"""Load the configuration data for custom tables.
+	
+	The expected data is a dictionary, typically loaded from a ConfigObj .ini file, of the form:
+	{
+		"filename1.utb": <TABLE_CONFIGURATION>,
+		"filename2.utb": <TABLE_CONFIGURATION>,
+		...
+	}
+	
+	Expected TABLE_CONFIGURATION is of the form:
+	{
+		"contracted": <bool>,  # Optional, defaults to False
+		"output": <bool>,      # Optional, defaults to True
+		"input": <bool>,       # Optional, defaults to True
+		"displayName": <DISPLAY_NAME>
+	}
+	
+	DISPLAY_NAME, if specified, can either be a string literal or a dictionary
+	mapping language or country codes to string literals.
+	If omitted, the display name will be the file name.
+	"""
+	for fileName, tableCfg in data.items():
+		addTable(
+			fileName,
+			displayName=_getCustomTableDisplayName(fileName, tableCfg),
+			contracted=tableCfg.get("contracted", False),
+			output=tableCfg.get("output", True),
+			input=tableCfg.get("input", True)
+		)
+
+
+def _getCustomTableDisplayName(fileName: str, tableCfg: dict) -> str:

```

Comment:
Have you considered making this a property on CustomTablesManifest?

Reply:
Actually, no.
The manifest is only one way to supply the config.
Another is to pass a dictionary to `brailleTables.loadCustomTablesConfig` (made public as of b87f5b22, see https://github.com/nvaccess/nvda/pull/10172#issuecomment-529537805)